---
title: "Differential Expression Analysis"
author: "Tiffany Tang"
date: "2025-07-17"
output: html_document
---

## Differential Expression Analysis code

In this markdown notebook we will be performing a differential expression analysis on our genes. 


```{r readin}
library(dplyr) # load necessary libraries
library(readr)

# Adjust path variables as needed
# pathBrian <- 
pathTiffany <- "C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Transcriptomics\\Final\\transformed_data.tsv"

# load in my data
# Takes a while to load in
expr_data <- read_tsv(pathTiffany)
```

## Analysis using limma

Since these are not the raw counts (they have been normalized before processing), we will be using limma to analyze the data.
- DESeq2 requires raw counts

With limma I want the expression matrix to be: 
- rows = genes/transcripts
- columns = samples

So I must reformat the data before this step.

```{r limma_reshape}

# Extract expression values 
# First 3 cols are sample, ID, type, so disregard those

expr_matrix <- as.matrix(expr_data[, 4:ncol(expr_data)])

rownames(expr_matrix) <- expr_data$ID  # sample IDs as column names

expr_matrix <- t(expr_matrix)  # Now rows = genes, cols = samples

head(expr_matrix) # just to look
# It looks alright

# Alright lets do the meta data
group <- factor(expr_data$type)  # type
design <- model.matrix(~0 + group)  # Model matrix for limma
# +0 removes the intercept, so we get a separate column for each group
design # Just to take a look, looking alright
```

Time to fit the model.

```{r limma_model}
# load in necessary libraries
library(edgeR)
library(limma)

#used the transposed matrix from earlier
fit <- lmFit(expr_matrix, design)

# Define contrasts, this is based on the groups in the design matrix
contrast_matrix <- makeContrasts(
  Lung_Cancer_vs_Normal = groupLung_Cancer - groupNormal_lung,
  levels = design
)


# Fit contrast
fit2 <- contrasts.fit(fit, contrast_matrix)
fit2 <- eBayes(fit2)

# Get top DE genes
results <- topTable(fit2, adjust="BH", number=Inf)

# View results
head(results)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

## Significant genes

Next step is to filter for significant genes that have an adjusted p-value of < 0.05 and a logFC > 1 or < -1

```{r sig_genes}

# Filter
sig_genes <- results %>%
  filter(adj.P.Val < 0.05 & abs(logFC) > 1)

# Check how many significant genes
nrow(sig_genes)

# Take a look
head(sig_genes)
```
There are 2313 significant genes. 

Create a subset of the expression matrix where it is only the significant genes.

```{r sig_matrix}
# Subset the expression matrix to only include significant genes
sig_ids <- rownames(sig_genes)

# Make sure gene IDs in expr_matrix are rownames
sig_expr_matrix <- expr_matrix[rownames(expr_matrix) %in% sig_ids, ]

# Add first column name

head(sig_expr_matrix)
```

Great! 


Lets write this out to a file and save it. 

```{r save_sig_genes}
library(tibble) 
# I keep losing the rownames so I need to save them as a column
sig_expr_df <- as.data.frame(sig_expr_matrix) %>%
  rownames_to_column(var = "GeneID")

# Please add the path to your variable below
# outpathBrian <- 
outpathTiffany <- "C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Transcriptomics\\Final\\significant_genes_limma.csv"

# Save the significant genes matrix
write_tsv(sig_expr_df, outpathTiffany)

```


