---
title: "Plotting"
author: "Tiffany Tang"
date: "2025-07-23"
output: html_document
---

## Code for visuals

This markdown notebook will have plots for clustering, heat maps and PCA.

Lets start by reading in our data. 
We will start by visualizing the significant gene data, and then compare this to the Transcription Factor only genes.

```{r readin}
library(dplyr) # load necessary libraries
library(readr)

# Adjust path variables as needed
# pathBrian <- 
pathTiffany <- "C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Transcriptomics\\Final\\significant_genes_limma.csv"

# load in my data
# This is for the significant genes!
sig_gene_data <- read_tsv(pathTiffany)

head(sig_gene_data)
```

## Clustering

We will be clustering and producing a heatmap

```{r heatmap}
# Load required library
library(pheatmap)

# Convert GeneID column to rownames
sig_gene_mat <- sig_gene_data
rownames(sig_gene_mat) <- sig_gene_mat$GeneID
sig_gene_mat <- sig_gene_mat[, -which(names(sig_gene_mat) == "GeneID")]

# Convert to matrix for heatmap as pheatmap requires matrix
sig_gene_mat <- as.matrix(sig_gene_mat)

# Scale the data 
sig_gene_mat_scaled <- t(scale(t(sig_gene_mat)))

# Generate clustered heatmap
pheatmap(sig_gene_mat_scaled,
         show_rownames = FALSE,     # Hide gene names because there are too many
         show_colnames = TRUE,      # Show sample names
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         fontsize_row = 6,
         fontsize_col = 10,
         main = "Clustering of Significant Genes")

```

We can see based on the heat map that most of the genes are less expressed than its average in the samples. 

I want to add labels for Cancer vs normal on this heat map. To do so I will load in the original expr_data and extract that meta data from there

```{r meta_data}
# Adjust path variables as needed
# pathBrian <- 
pathTiffany <- "C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Transcriptomics\\Final\\transformed_data.tsv"

# load in my data
expr_data <- read_tsv(pathTiffany)

head(expr_data)

# list of significant samples
sig_samples <- colnames(sig_gene_data)[-1]

# Subset expr_data to only rows used in heatmap
annotation_df <- as.data.frame(expr_data[expr_data$ID %in% sig_samples, c("ID", "type")])

# Make the sample IDs as the rownames
rownames(annotation_df) <- annotation_df$ID

# Remove the redundant row
annotation_df <- annotation_df[, "type", drop = FALSE]

# Create another column in this that lets us know if it is a cancer sample or not
# annotate the heat map based on this
annotation_df$CancerStatus <- ifelse(grepl("cancer", annotation_df$type, ignore.case = TRUE),
                                     "Cancer", "Normal")

```

Now we have an annotation dataframe, lets re run the heatmap code

```{r annotate_heatmap}
pheatmap(sig_gene_mat_scaled,
         show_rownames = FALSE,
         show_colnames = TRUE,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         fontsize_row = 6,
         fontsize_col = 10,
         annotation_col = annotation_df,
         main = "Clustering of Significant Genes by Cancer Status")

```

Now we can see that most of the cancer samples have lower expression than the normal samples!

What if we crossed this with the Transcription Factor genes?

```{r tf_sig_heatmap}
# read in tf only list
# Adjust path variables as needed
# pathBrian <- 
pathTiffany <- "C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Transcriptomics\\Final\\tfOnly_list.txt"

# load in my data
tf_list <- read_tsv(pathTiffany, col_names = FALSE)
#Make it a vector
tf_list <- tf_list[[1]]


# Isolate significant gene IDs
gene_ids <- sig_gene_data$GeneID

# Create a logical vector indicating TF membership
is_tf <- gene_ids %in% tf_list

#Create dataframe to mark TF status
tf_status <- as.data.frame(ifelse(is_tf, "TF", "Non-TF"))

rownames(tf_status) <- gene_ids
colnames(tf_status) <- "TF_Status"

# Heat map plot code
pheatmap(sig_gene_mat_scaled,
         show_rownames = TRUE,
         show_colnames = FALSE,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         fontsize_row = 6,
         fontsize_col = 10,
         annotation_row = tf_status,
         main = "Clustering of Significant Genes")

```
**I cannot seem to resolve this issue even though nothing is wrong**

We will move on from this. 

Plot with only significant genes that are transcription factors. 

```{r sig_TF_heatmap}
# Subset the significant gene data to TF-only rows
sig_TF_data <- as.data.frame(sig_gene_data %>% filter(GeneID %in% tf_list))

# Set rownames and drop GeneID column
rownames(sig_TF_data) <- sig_TF_data$GeneID
sig_TF_data <- sig_TF_data[, -which(names(sig_TF_data) == "GeneID")]

# Convert to matrix
sig_TF_mat <- as.matrix(sig_TF_data)

# Scale the rows (genes)
sig_TF_mat_scaled <- t(scale(t(sig_TF_mat)))

# Create annotation for samples 
# Make sure it includes only the relevant samples
annotation_df_TF <- annotation_df[colnames(sig_TF_mat_scaled), , drop = FALSE]

# Plot the heatmap
pheatmap(sig_TF_mat_scaled,
         show_rownames = FALSE,
         show_colnames = TRUE,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         fontsize_row = 6,
         fontsize_col = 10,
         annotation_col = annotation_df_TF,
         main = "Clustering of Significant Transcription Factor Genes")

```
We can see that the trend observed earlier: lower expression in cancer samples compared to normal samplesâ€”remains evident. This pattern appears to be even more pronounced when focusing specifically on the transcription factor genes.


## PCA Plot

Now we will create the PCA analysis and its corresponding plot. 

```{r sig_genes_pca}
# set up data for PCA
sig_pca <- sig_gene_data

# Set col gene ID to the row names and dete
rownames(sig_pca) <- sig_pca$GeneID
sig_pca <- sig_pca[, -which(colnames(sig_pca)=="GeneID")]

sig_gene_pca_input<- t(sig_pca)

# Perform PCA
pca_res <- prcomp(sig_gene_pca_input, scale. = FALSE)

# Make data frame for plotting
pca_df <- as.data.frame(pca_res$x)
pca_df$SampleID <- rownames(pca_df)

# Add CancerStatus from annotation_df by rownames
pca_df$CancerStatus <- annotation_df[pca_df$SampleID, "CancerStatus"]

# Plot
library(ggplot2)

ggplot(pca_df, aes(x = PC1, y = PC2, color = CancerStatus)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "PCA of Significant Genes",
       x = paste0("PC1 (", round(summary(pca_res)$importance[2,1]*100, 1), "%)"),
       y = paste0("PC2 (", round(summary(pca_res)$importance[2,2]*100, 1), "%)"))

```
PCA plot with Cancer type
```{r pca_cancertype}
# Add CancerStatus from annotation_df by rownames
pca_df$CancerType <- annotation_df[pca_df$SampleID, "type"]

# Plot
library(ggplot2)

ggplot(pca_df, aes(x = PC1, y = PC2, color = CancerType)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "PCA of Significant Genes",
       x = paste0("PC1 (", round(summary(pca_res)$importance[2,1]*100, 1), "%)"),
       y = paste0("PC2 (", round(summary(pca_res)$importance[2,2]*100, 1), "%)"))
```

PCA of significant genes that are transcription factors. 

```{r sig_TF_pca}
library(ggplot2)

# Set data, make sure it is only the significant genes that are also transcription factors!
sig_tf_pca <- as.data.frame(sig_gene_data %>% filter(GeneID %in% tf_list))

# Set rownames and drop GeneID column
rownames(sig_tf_pca) <- sig_tf_pca$GeneID

# This is an expressiong df of just genes that are transcription factors
sig_tf_pca <- sig_tf_pca[, -which(names(sig_tf_pca) == "GeneID")]

sig_tf_pca <- t(sig_tf_pca)  # Samples need to be rows

# PCA
pca_result <- prcomp(sig_tf_pca, center = TRUE, scale. = TRUE)

#Make it a dataframe
pca_df <- as.data.frame(pca_result$x) 
pca_df$SampleID <- rownames(pca_df) # Add some rownames
# Add CancerStatus from annotation_df by rownames
pca_df$CancerType <- annotation_df[pca_df$SampleID, "type"]


# Plot
ggplot(pca_df, aes(x = PC1, y = PC2, color = CancerType)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "PCA of Significant Genes That are Transcription Factors",
       x = paste0("PC1 (", round(summary(pca_result)$importance[2,1]*100, 1), "%)"),
       y = paste0("PC2 (", round(summary(pca_result)$importance[2,2]*100, 1), "%)"))
```









































