---
title: "Plotting"
author: "Tiffany Tang"
date: "2025-07-23"
output: html_document
---

## Code for visuals

This markdown notebook will have plots for clustering, heat maps and PCA.

Lets start by reading in our data. 
We will start by visualizing the significant gene data, and then compare this to the Transcription Factor only genes.

```{r readin}
library(dplyr) # load necessary libraries
library(readr)

# Adjust path variables as needed
# pathBrian <- 
pathTiffany <- "C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Transcriptomics\\Final\\significant_genes_limma.csv"

# load in my data
# This is for the significant genes!
sig_gene_data <- read_tsv(pathTiffany)

head(sig_gene_data)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
